### PR TITLE
[Ubuntu] Slight rework of the PyPy installation

### PR DESCRIPTION
**Pull request by shilovmaksim**
# Description
Avoid parsing PyPy downloads "html" page. Use versions config provided by PyPy developers instead, which is considered to be more stable.

## Check list
- [x] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ ] Documentation is updated (if applicable)
- [x] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/linux/scripts/installers/pypy.sh [hunk 1]
```diff
@@ -74,9 +74,8 @@ function InstallPyPy
 }
 
 # Installation PyPy
-uri="https://downloads.python.org/pypy/"
-download_with_retries $uri "/tmp" "pypyUrls.html" compressed
-pypyVersions="$(cat /tmp/pypyUrls.html | grep 'linux64' | awk -v uri="$uri" -F'>|<' '{print uri$5}')"
+uri="https://downloads.python.org/pypy/versions.json"
+pypyVersions=$(curl $uri | jq '.[].files[].download_url' | grep 'linux64')
```
[the remark this answers is not in the document]
Comment (shilovmaksim) [reply images/linux/scripts/installers/pypy.sh]: Done

File: images/linux/scripts/installers/pypy.sh [hunk 2]
```diff
@@ -74,14 +74,12 @@ function InstallPyPy
...
-download_with_retries $uri "/tmp" "pypyUrls.html" compressed
-pypyVersions="$(cat /tmp/pypyUrls.html | grep 'linux64' | awk -v uri="$uri" -F'>|<' '{print uri$5}')"
+pypyVersions=$(curl https://downloads.python.org/pypy/versions.json)
 toolsetVersions=$(get_toolset_value '.toolcache[] | select(.name | contains("PyPy")) | .versions[]')
 
 for toolsetVersion in $toolsetVersions; do
-    latestMajorPyPyVersion=$(echo "${pypyVersions}" | grep -E "pypy${toolsetVersion}-v[0-9]+\.[0-9]+\.[0-9]+-" | head -1)
-
+    latestMajorPyPyVersion=$(echo $pypyVersions | jq --arg toolsetVersion $toolsetVersion '.[] | select((.python_version | contains($toolsetVersion)) and .stable == true).files[] | select(.arch == "x64" and .platform == "linux").download_url' | head -1)
+    latestMajorPyPyVersion=`sed -e 's/^"//' -e 's/"$//' <<<"$latestMajorPyPyVersion"`
```
Comment (al-cheb) [inline images/linux/scripts/installers/pypy.sh:82]: remove this command and add  a param -r to jq

File: images/linux/scripts/installers/pypy.sh [hunk 3]
```diff
@@ -74,14 +74,12 @@ function InstallPyPy
...
-uri="https://downloads.python.org/pypy/"
-download_with_retries $uri "/tmp" "pypyUrls.html" compressed
-pypyVersions="$(cat /tmp/pypyUrls.html | grep 'linux64' | awk -v uri="$uri" -F'>|<' '{print uri$5}')"
+pypyVersions=$(curl https://downloads.python.org/pypy/versions.json)
 toolsetVersions=$(get_toolset_value '.toolcache[] | select(.name | contains("PyPy")) | .versions[]')
 
 for toolsetVersion in $toolsetVersions; do
-    latestMajorPyPyVersion=$(echo "${pypyVersions}" | grep -E "pypy${toolsetVersion}-v[0-9]+\.[0-9]+\.[0-9]+-" | head -1)
-
+    latestMajorPyPyVersion=$(echo $pypyVersions | jq --arg toolsetVersion $toolsetVersion '.[] | select((.python_version | contains($toolsetVersion)) and .stable == true).files[] | select(.arch == "x64" and .platform == "linux").download_url' | head -1)
```
Comment (al-cheb) [inline images/linux/scripts/installers/pypy.sh:81]: This filter is quite long, I think we should split it on two parts:
1. Get all download_url
2. Use grep to take first occurrence of $toolsetVersion

File: images/linux/scripts/installers/pypy.sh (hunk 2 above)
Comment (shilovmaksim) [reply images/linux/scripts/installers/pypy.sh:82]: Maybe I'm doing smth wrong but "-r" not working locally
`jq: error: Could not open file .[] | select((.python_version | contains($toolsetVersion)) and .stable == true).files[] | select(.arch == "x64" and .platform == "linux").download_url: No such file or directory`

File: images/linux/scripts/installers/pypy.sh (hunk 3 above)
Comment (shilovmaksim) [reply images/linux/scripts/installers/pypy.sh:81]: jq supports line splitting. How about just split it to smth like this:
```
latestMajorPyPyVersion=$(echo $pypyVersions |
    jq --arg toolsetVersion $toolsetVersion '.[]
    | select((.python_version | contains($toolsetVersion)) and .stable == true).files[]
    | select(.arch == "x64" and .platform == "linux").download_url' | head -1)
```

File: images/linux/scripts/installers/pypy.sh (hunk 2 above)
Comment (al-cheb) [reply images/linux/scripts/installers/pypy.sh:82]:
```
:~$ echo $pypyVersions | jq -r --arg toolsetVersion $toolsetVersion '.[] | select((.python_version | contains($toolsetVersion)) and .stable == true).files[] | select(.arch == "x64" and .platform == "linux").download_url' | head -1
https://downloads.python.org/pypy/pypy3.8-v7.3.7-linux64.tar.bz2
```

Comment (shilovmaksim) [reply images/linux/scripts/installers/pypy.sh]: Updated

File: images/linux/scripts/installers/pypy.sh (hunk 3 above)
Comment (shilovmaksim) [reply images/linux/scripts/installers/pypy.sh]: Updated

File: images/linux/scripts/installers/pypy.sh [hunk 4]
```diff
@@ -74,14 +74,14 @@ function InstallPyPy
...
-pypyVersions="$(cat /tmp/pypyUrls.html | grep 'linux64' | awk -v uri="$uri" -F'>|<' '{print uri$5}')"
+pypyVersions=$(curl https://downloads.python.org/pypy/versions.json)
 toolsetVersions=$(get_toolset_value '.toolcache[] | select(.name | contains("PyPy")) | .versions[]')
 
 for toolsetVersion in $toolsetVersions; do
-    latestMajorPyPyVersion=$(echo "${pypyVersions}" | grep -E "pypy${toolsetVersion}-v[0-9]+\.[0-9]+\.[0-9]+-" | head -1)
-
+    latestMajorPyPyVersion=$(echo $pypyVersions |
+        jq -r --arg toolsetVersion $toolsetVersion '.[]
+        | select((.python_version | contains($toolsetVersion)) and .stable == true).files[]
```
Comment (miketimofeev) [inline images/linux/scripts/installers/pypy.sh:83]: nit, to safe on the safe side:
```suggestion
        | select((.python_version | startswith($toolsetVersion)) and .stable == true).files[]
```

Comment (shilovmaksim) [reply images/linux/scripts/installers/pypy.sh:83]: Indeed

**Review by al-cheb — APPROVED** [review]

**Review by miketimofeev — APPROVED** [review]

**Review by miketimofeev — APPROVED** [review]

**Pull request merged**